### PR TITLE
Add XDP support and TCP Syn flooding into the attack tool.

### PR DESCRIPTION
**Pull request by VipAlekseyPetrenko**
## Description
Enabled XDP support in the attack tool to allow for faster flooding rates by the attackers.
Enabled TCP Syn flooding attack.
Added a helper function for checking if RawDataPath is enabled on the CXPLAT_DATAPATH variables.

## Testing
Manually tested my changes with 2 attacker VMs (one with XDP enabled and one with XDP disabled) and having a Windows Server running with a WireShark tool reading incoming packets.

## Documentation
No.

### CODE REVIEW

**Comment by VipAlekseyPetrenko** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: src/tools/attack/attack.cpp [hunk 1]
```diff
@@ -106,31 +101,55 @@ UdpUnreachCallback(
...
-    while (CxPlatTimeDiff64(TimeStart, CxPlatTimeMs64()) < TimeoutMs) {
+    CXPLAT_ROUTE Route = {0};
+    CxPlatSocketGetLocalAddress(Binding, &Route.LocalAddress);
+    CxPlatSocketGetRemoteAddress(Binding, &Route.RemoteAddress);
 
-        CXPLAT_SEND_CONFIG SendConfig = { &Route, Length, CXPLAT_ECN_NON_ECT, 0 };
-        CXPLAT_SEND_DATA* SendData = CxPlatSendDataAlloc(Binding, &SendConfig);
-        if (SendData == nullptr) {
-            printf("CxPlatSendDataAlloc failed\n");
+    if (CxPlatIsRawInitialized(Datapath))
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:133]: `CxPlatIsRawInitialized` should be unnecessary. Just always call this function, and it's a no-op if not using raw.

File: src/tools/attack/attack.cpp [hunk 2]
```diff
@@ -378,82 +385,72 @@ main(
     _In_reads_(argc) _Null_terminated_ char* argv[]
     )
 {
-    int ErrorCode = -1;
+    if (argc < 2) {
+        PrintUsage();
+        return -1;
+    }
+    else if (strcmp("-list", argv[1]) == 0) {
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:392]:
```suggestion
    } else if (strcmp("-list", argv[1]) == 0) {
```

File: src/tools/attack/attack.cpp [hunk 3]
```diff
@@ -378,82 +385,72 @@ main(
...
+        PrintUsageList();
+        return 0;
+    }
+    else if (!TryGetValue(argc, argv, "type", &AttackType) ||
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:396]:
```suggestion
    } else if (!TryGetValue(argc, argv, "type", &AttackType) ||
```

File: src/tools/attack/attack.cpp [hunk 4]
```diff
@@ -378,82 +385,72 @@ main(
...
+    (AttackType < 1 || AttackType > 4)) {
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:397]:
```suggestion
        (AttackType < 1 || AttackType > 4)) {
```

File: src/tools/attack/attack.cpp [hunk 5]
```diff
@@ -378,82 +385,72 @@ main(
...
+    RunAttack();
 
     CxPlatDataPathUninitialize(Datapath);
     CxPlatUninitialize();
     CxPlatSystemUnload();
-
-    return ErrorCode;
-}
+    return 0;
+}
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:456]: Please make sure all files have a trailing (empty) new line

File: src/tools/attack/attack.cpp [hunk 6]
```diff
@@ -13,24 +13,18 @@
 
 #include "precomp.h" // from 'core' dir
 #include "msquichelper.h"
-
 #include "packet_writer.h"
 
+//========================================Vars/Consts========================================
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:18]: Let's remove these big comments please.

File: src/tools/attack/attack.cpp [hunk 7]
```diff
@@ -106,31 +101,55 @@ UdpUnreachCallback(
...
+    CxPlatSocketGetLocalAddress(Binding, &Route.LocalAddress);
+    CxPlatSocketGetRemoteAddress(Binding, &Route.RemoteAddress);
 
-        CXPLAT_SEND_CONFIG SendConfig = { &Route, Length, CXPLAT_ECN_NON_ECT, 0 };
-        CXPLAT_SEND_DATA* SendData = CxPlatSendDataAlloc(Binding, &SendConfig);
-        if (SendData == nullptr) {
-            printf("CxPlatSendDataAlloc failed\n");
+    if (CxPlatIsRawInitialized(Datapath))
+    {
+        if (QUIC_FAILED(CxPlatResolveRoute(Binding, &Route, 0, &Route, ResolveRouteComplete))) {
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:135]: This is asynchronous. You need to wait on it.

File: src/tools/attack/attack.cpp (hunk 1 above)
Comment (anrossi) [reply src/tools/attack/attack.cpp:133]: I think Alex was seeing a crash here when using the TCP test if XDP wasn't installed on the system. Basically the automatic fallback wasn't happening.

Comment (VipAlekseyPetrenko) [reply src/tools/attack/attack.cpp]: Good point, changed.

File: src/tools/attack/attack.cpp (hunk 2 above)
Comment (VipAlekseyPetrenko) [reply src/tools/attack/attack.cpp]: Fixed.

File: src/tools/attack/attack.cpp (hunk 3 above)
Comment (VipAlekseyPetrenko) [reply src/tools/attack/attack.cpp]: Fixed.

File: src/tools/attack/attack.cpp (hunk 4 above)
Comment (VipAlekseyPetrenko) [reply src/tools/attack/attack.cpp]: Fixed.

File: src/tools/attack/attack.cpp (hunk 5 above)
Comment (VipAlekseyPetrenko) [reply src/tools/attack/attack.cpp]: Done.

File: src/tools/attack/attack.cpp (hunk 6 above)
Comment (VipAlekseyPetrenko) [reply src/tools/attack/attack.cpp]: Done.

File: src/tools/attack/attack.cpp (hunk 7 above)
Comment (VipAlekseyPetrenko) [reply src/tools/attack/attack.cpp]: Done.

**Comment by VipAlekseyPetrenko** [comment]
Acted on comments.

File: src/tools/attack/attack.cpp [hunk 8]
```diff
@@ -106,31 +125,57 @@ UdpUnreachCallback(
...
 {
     CXPLAT_ROUTE Route = {0};
     CxPlatSocketGetLocalAddress(Binding, &Route.LocalAddress);
-    Route.RemoteAddress = ServerAddress;
+    CxPlatSocketGetRemoteAddress(Binding, &Route.RemoteAddress);
+    CallbackContext Context = {&Route, };
+    QUIC_STATUS Status = CxPlatResolveRoute(Binding, &Route, 0, &Context, ResolveRouteComplete);
+    if (Status == QUIC_STATUS_PENDING) {
+        CxPlatEventInitialize(Context.Event, FALSE, FALSE);
+        if (!CxPlatEventWaitWithTimeout(Context.Event, (DWORD)TimeoutMs)) {
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:154]: You're not cleaning up the event if you timeout.

File: src/inc/quic_datapath.h [hunk 9]
```diff
@@ -770,6 +770,12 @@ CxPlatUpdateRoute(
     _In_ CXPLAT_ROUTE* SrcRoute
     );
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+BOOLEAN
+CxPlatIsRawInitialized(
+    _In_ CXPLAT_DATAPATH* Datapath
+    );
+
```
[the remark this answers is not in the document]
Comment (VipAlekseyPetrenko) [reply src/inc/quic_datapath.h:778]: Got it, fixed.

File: src/tools/attack/attack.cpp (hunk 8 above)
Comment (VipAlekseyPetrenko) [reply src/tools/attack/attack.cpp:154]: Fixed.

**Review by nibanks — DISMISSED** [review]

**Pull request merged**